### ISSUE TITLE
Scriptable options for dataset properties invoked before element initialization

### ISSUE THREAD

**Issue by stockiNail** [kurkle/chartjs-chart-treemap#120]
During the update, the controller is using some dataset options which are not leveraging on the treemap element (because not initialized), for the scriptable options.

At the moment, these properties are mainly used to calculate the space for the caption and can not be defined as callback.

I think these options should be consumed even if they are defined as scriptable options.

The context should slightly different form the data context and maybe a specific context object should be implemented.

The following are the options in scope:

| options
| :-
| `dataset.spacing`
| `dataset.borderWidth`
| `dataset.borderRadius` (probably, when PR #114 will be merged, needed to improve the sub-rect calculation)
| `dataset.captions.font`
| `dataset.captions.padding`

**Comment by kurkle** [comment]
The context type is not different?

**Comment by stockiNail** [comment]
With your sentence, I'm thinking that I'm completely wrong... I remembered that I don't have any context in the dataset therefore, if we want to invok the callbacks (by the function from CHART.JS helpers) I need to create the context with their properties.... for instance, `raw` option (mandatory if you want to leverage on the data to apply a specific logic) Am I wrong?

**Comment by kurkle** [comment]
The dataset controller instance options property should be a resolver with 'dataset' context. The dataset options should be read from that options object (we are reading from dataset directly, right?)

**Comment by stockiNail** [comment]
yes, correct. But I don't have any raw data to allow the use to define, for instance, different font for different groups. Do you think we can update it? the issue is to set the raw data in a context which is unique for the dataset. Therefore, to keep it clean, we should add raw data and remove it afterwards.

**Comment by stockiNail** [comment]
At the first call, the dataset is pure options passed in the chart config, no context at all.

The controller is using `getDataset` method:

```javascript
  getDataset() {
    return this.chart.data.datasets[this.index];
  }
```

**Comment by stockiNail** [comment]
I'm investigating because `buildData` is invoked with the element as well, therefore the space calculation could be done only when a context is available.

**Comment by kurkle** [comment]
> I'm investigating because `buildData` is invoked with the element as well, therefore the space calculation could be done only when a context is available.

I have the dataset -> options almost ready, will submit soon.

**Comment by stockiNail** [comment]
Closed in favor of #122
